Multi-field search, improve data storage

mpuckett159 opened the pull request:
Closes #17  and #19 

This adds the sqlite database backend, which is created before the app is pushed to whatever hosting site it will live on and doesn't require the it to be loaded to on application launch every time.

This also allows for the multi-field search now that the SQL backend is set up.

This all maintains the older functionality as well.

mpuckett159 committed:
Add multiple search fields

app.py:
Update context objects to provide an array of entities for the HTML form to use.

Update page functions to utilize dataset HTML page rendering function handling no queries passed.

dataset.py:
Add if/else to handle empty queries so we can clean up the app.py route functions to be simpler.

index.html:
Update HTML form to loop through form entities to support multiple search terms.

mpuckett159 committed:
Ignore DS_Store files

mpuckett159 committed:
Clean up data files

Clean up existing CSV file

Create 1-312-data sqlite DB file. It's a binary so not ideal to store in git but doing this so it can be pulled into the Glitch app without having to waste cycles reading the CSV into the database. We will leave the CSV for now.

mpuckett159 committed:
Update app to utilize sqlite3 db

app.py:
Remove vehicle make loop up entity from entities list. This was added to test some stuff but I don't want to add this now, though it could be added later.

Remove BADGE_CONTEXT because it is no longer needed.

Remove /badge route because it is no longer needed

Move /badge-lookup/<badge> route to the end of the file for archiving old stuff for maintaining compatibility.

Update /name route to properly return a None value for param gets so that the SQL queries don't get fucked up by blank fields.

Update name_lookup to use new inputs

Create archive section for old endpoint compatibility maintenance.

Update the old endpoints to use the new input requirements for the name_lookup function

Remove BADGE_CONTEXT usage again.

dataset.py:
Remove unneeded modules

Add sqlite3 module to use new database

Remove CSV reading stuff.

Connecting to the sqlite3 database and storing the cursor

Removing the supporting function and class to load the CSV data

Update _augment_with_salary to use new SQL returns, which are Tuples.

Remove badge_lookup function

Update _sort_names function to use new SQL returns, which are Tuples.

Update name_lookup function to utilize the SQL queries.

index.html:
Update form override buttons to be Vehicles and Officers from Licenses and Names. This more closely aligns with what the buttons load. I also removed the Badges button because it is unnecessary imo since I've added the ability to search multiple fields for the Names context.

AetherUnbound commented on src/app.py:
```diff
@@ -35,78 +39,68 @@ def license_page():
-    html = ""
-    if last_name:
-        html = dataset.name_lookup(last_name)
+    last_name = None if request.args.get('last') == "" else request.args.get('last')
+    first_name = None if request.args.get('first') == "" else request.args.get('first')
+    badge = None if request.args.get('badge') == "" else request.args.get('badge')
```
For each of these, I think we could use the following logic to simplify this:
```suggestion
    last_name = request.args.get('last') or None
    first_name = request.args.get('first') or None
    badge = request.args.get('badge') or None
```

Ex:
```
[ins] In [6]: def assign(x):
         ...:     print(x or None)
         ...: 

[ins] In [7]: assign("foo")
foo

[ins] In [8]: assign("")
None
```

Do these need to be `None`? Is there a reason they can't be an empty string?

mpuckett159 replied:
Brief explanation for why they need to be `None`:

```
records = sql_curs.execute(
                "SELECT * FROM officers WHERE Last = ? OR First = ? OR Serial = ?",
                (last_name, first_name, badge),
            ).fetchall()
```

You can see here that I'm using the sqlite library's cursor object, and doing a safe injection of the values to filter the query on from the `(last_name, first_name, badge)` tuple into the `?`s in order. The way the sqlite library protects against SQL injection attacks here is it wraps any strings injected in `"`s so that you can't escape out, so, when you pass any string object, even if it's empty, it gets wrapped in quotes, so you end up with:

`SELECT * FROM officers WHERE Last = "" OR First = "" OR Serial = ""`

Which will match empty cells instead of skipping the column. When the variables are just straight up `None` (as opposed to `str(None)`), you get this I think:

`SELECT * FROM officers WHERE Last =  OR First =  OR Serial = `

Which will just skip the column searches since no value is passed.

This explanation is provided for completeness we already discussed elsewhere.

AetherUnbound commented on src/dataset.py:
```diff
@@ -98,43 +69,33 @@ def _augment_with_salary(record: Dict[str, str]) -> Dict[str, str]:
-
-def _sort_names(record: _DataDict):
-    for r in sorted(record.values(), key=lambda x: x[C.FIRST]):
+def _sort_names(record: List[Tuple]) -> Tuple:
```
I might suggest moving this function up before the augment, and instead converting the record at this point to a `NamedTuple`. This would make access to the return values a little bit clearer (also note I've changed the variable name and the return type):

```python
class RosterRecord(NamedTuple):
    serial: int
    first: str
    last: str
    middle: str
    title: str
    unit: str

# Usage
def _sort_names(records: List[Tuple]) -> Generator[RosterRecord, None, None]: 
    # Convert to named tuples
    records = [RosterRecord(*r) for r in records]
    for r in sorted(records, key=lambda x: x.last): 
        yield r


# Later usage
record.title
record.unit
# Etc
```

AetherUnbound commented on src/dataset.py:
```diff
@@ -1,87 +1,58 @@
-    first = record[C.FIRST]
-    last = record[C.LAST]
-    middle = record[C.MIDDLE]
+def _augment_with_salary(record: Tuple) -> Dict[str, str]:
```
With the change suggested to the function below, this would be `record: RosterRecord`, and the values could be accessed as attributes rather than tuple indices.

AetherUnbound commented on src/dataset.py:
```diff
@@ -98,43 +69,33 @@ def _augment_with_salary(record: Dict[str, str]) -> Dict[str, str]:
+    if last_name or first_name or badge:
+        try:
+            records = sql_curs.execute(
+                "SELECT * FROM officers WHERE Last = ? OR First = ? OR Serial = ?",
```
I would suggest using `LIKE` rather than a hard match - this will allow users to make close approximation searches (most DBs have `ILIKE` for case insensitive like, but it seems `LIKE` is [case insensitive in sqlite](https://www.sqlitetutorial.net/sqlite-like/))
According to [this SO question](https://stackoverflow.com/a/3105370/3277713), you'll have to add the wildcards (`%`) to the string prior to executing the query. We won't want to do this to `None` values that are supplied to the function because I think that will make it match all records lol.

mpuckett159 replied:
Yeah I need to play around with the SQL query options more. I'm going to be updating this to be a little more dynamic as well, and doing something similar to:

```
>>> sql_query = "SELECT * FROM officers WHERE Last = ?"
>>> sql_curs.execute(sql_query, ("Aagard",),).fetchall()
[('5011', 'Lori', 'K.', 'Aagard', 'Police Lieutenant', 'South Pct Ops')]
>>> sql_query += " OR First = ?"
>>> sql_query
'SELECT * FROM officers WHERE Last = ? OR First = ?'
>>> sql_curs.execute(sql_query, ("Aagard","Lori",),).fetchall()
[('5011', 'Lori', 'K.', 'Aagard', 'Police Lieutenant', 'South Pct Ops'), ('5819', 'Lori', 'A', 'Nelsen-Allen', 'Admin Specialist I', 'Dv - Domestic Violence Unit - Admi~')]
```

I didn't realize you could do this at first but it makes a lot more sense and also will simplify the query param gets above so that it doesn't matter if they're `str(None)` or just `None`.

AetherUnbound replied:
Yea it's a very common pattern to have the base SQL statement and then build up the `WHERE` clauses as needed with different parameters!

AetherUnbound commented on src/dataset.py:
```diff
@@ -98,43 +69,33 @@ def _augment_with_salary(record: Dict[str, str]) -> Dict[str, str]:
-
-    print("Final HTML:\n{}".format(html))
-    return html
+def name_lookup(last_name: Any, first_name: Any, badge: Any) -> str:
```
FWIW this would be `Optional[str]` rather than `Any`

mpuckett159 replied:
This will be updated to just `str` now that I know how to properly build dynamic SQL queries.

AetherUnbound commented on src/dataset.py:
```diff
@@ -98,43 +69,33 @@ def _augment_with_salary(record: Dict[str, str]) -> Dict[str, str]:
-    print("Final HTML:\n{}".format(html))
-    return html
+def name_lookup(last_name: Any, first_name: Any, badge: Any) -> str:
+    if last_name or first_name or badge:
```
It might be better to invert this conditional and exit early for cleanliness rather than have most of the logic in this function in the if block here

AetherUnbound commented on src/views/index.html:
```diff
@@ -19,19 +19,18 @@ <h1>
     <p>Enter in a {{ entity_name_long }}</p>
     <form action="/{{ lookup_url }}" method="GET">
-        <input id="{{ query_param }}" name="{{ query_param }}" placeholder="{{ entity_name_short }}">
+        {% for entity in entities %}
```
Very nice!

AetherUnbound reviewed (commented):
This is an awesome start! I've added a couple of comments on things I think can be modified. 

I think it would also be extremely useful to include to include the DDLs (data definition language, essentially the `CREATE TABLE` statements) for the tables you made in the sqlite database in this PR, possibly under a "migrations" folder. It'll be necessary for historical purposes, to know what datatypes are present without cracking open the DB itself, and to know what fields may need to be migrated in the future should the data change.

mpuckett159 committed:
Applying comments, mostly

app.py:
 * Reorder entities so search fields appear as
    - First Name
    - Last Name
    - Badge Number

* Revert query param gets back to just simple `.get`s to work with new query building format.

dataset.py:
 * Update typing import to match reality.
 * Add table to explain database format
 * Add RosterRecord class to simplify working with database records
 * Update _sort_names to make use of new RosterRecord class
 * Broke up raw HTML string to conform to PEP8 or which ever the "no long lines" one is.
 * Update _augment_with_salary to make use of new RosterRecord Class
 * Update name_lookup function a couple ways which I will explain below

name_lookup was updated so that the order of the inputs for the function followed a more natural (to me at least) ordering. First, Last, Badge#. I've also moved the empty return to the top as suggested.

I tried simplifying the code a little bit so we aren't relying on a bunch of if statements to build queries. Let me know what you think on that _build_sql_query function. It could probably be simplified further, not sure.

AetherUnbound commented on src/dataset.py:
```diff
@@ -1,93 +1,94 @@
+sql_conn = sqlite3.connect('./data/1-312-data.db')
+sql_curs = sql_conn.cursor()
+
+#################################################################################
+# Database format
+#################################################################################
+# Example:
+# Table
+# - column (datatype)
+#################################################################################
+# officers
+# - Serial (string)
+# - First (string)
+# - Middle (string)
+# - Last (string)
+# - TitleDescription (string)
+# - UnitDescription (string)
+#################################################################################
```
This is good info, but I think we should have an actual `.sql` file with a `CREATE TABLE` statement for the sqlite table that we're using presently.

AetherUnbound commented on src/dataset.py:
```diff
@@ -98,43 +99,43 @@ def _augment_with_salary(record: Dict[str, str]) -> Dict[str, str]:
-        html = f"<p><b>Error:</b> {err}"
+def _build_sql_query(queries: List[Tuple]) -> (str, Tuple):
+    # (column, operator, value)
+    query_list = []
+    query_tuple = ()
+    for query in queries:
+        if query[2]:
+            query_list.append(" ".join(query[:2]) + " ?")
+            query_tuple += (query[2],)
+    return " AND ".join(query_list), query_tuple
```
I believe this will also work, and makes the code a bit clearer. Additionally, we want to be sure we wrap the `LIKE` parameters in wildcards so that users can make partial searches. I also think that the parameters can be any sequence, and don't need to be a tuple specifically.

```suggestion
    query_statements = []
    parameters = []
    for field, operator, value in queries:
        if not value:
            continue
        if operator == "LIKE":
            # Wrap query parameters in wildcards
            value = f"%{value}%"
        query_statements.append(f"{field} {operator} ?")
        parameters.append(value)
    return " AND ".join(query_statements), parameters
```

AetherUnbound reviewed (commented):
Looking so clean! Just two more suggestions before we merge

mpuckett159 committed:
Add wildcard usage info in page

This commit adds the wildcard usage info to the NAME_CONTEXT to make it clearer to users what they can do without being obtrusive.

mpuckett159 committed:
Add .sql file

Add .sql file to properly describe the database table.

mpuckett159 committed:
Update _build_sql_query to be better

Update _build_sql_query to simplify how we're building filters. The for loop declaration breaks up the query tuple into easier to understand variables instead of using numerical indexes.

AetherUnbound commented on src/app.py:
```diff
@@ -53,12 +53,12 @@ def license_page():
         {
             "entity_name_long": "first name",
-            "entity_name_short": "First Name",
+            "entity_name_short": "First Name - You can use % for a wildcard match",
```
```suggestion
            "entity_name_short": "First Name - Start or end name with % for partial matches",
```
Thoughts on wording (to make it a bit more accessible). Same for below

AetherUnbound commented on src/data/1-312-data.sql:
```diff
@@ -0,0 +1,8 @@
+CREATE TABLE IF NOT EXISTS officers (
```
:sparkles: :tada: Perfect! Do you mind moving this into a `migrations` subfolder, and adding the date to the filename? This helps for future migrations.

AetherUnbound commented on src/dataset.py:
```diff
@@ -100,14 +100,14 @@ def _augment_with_salary(record: RosterRecord) -> Dict[str, str]:
 
 
 def _build_sql_query(queries: List[Tuple]) -> (str, Tuple):
```
I think I would write this as 
```suggestion
def _build_sql_query(queries: List[Tuple]) -> Tuple[str, List[str]]:
```

AetherUnbound reviewed (commented):
Just a couple more minor things

mpuckett159 committed:
Some updates

app.py:
 * Update entity_name_short to help to make it easier to understand how to use wildcards.

12-20-2020_1-312-data.sql:
 * Move this file to src/data/migrations directory
 * Update file name to reflect date of migration

dataset.py:
 * Update _build_sql_query function to have an accurate return type helper.

AetherUnbound reviewed (approved):
Woohoo!

mpuckett159 committed:
Close sqlite connection properly

Properly close out the connection to the sqlite DB so as to avoid "not opened with the right thread" errors.

AetherUnbound commented on src/dataset.py:
```diff
@@ -123,15 +120,17 @@ def name_lookup(first_name: str, last_name: str, badge: str) -> str:
-                    context = _augment_with_salary(r)
-                    htmls.append(render_template("officer.j2", **context))
-                html = "\n<br/>\n".join(htmls)
+            with sqlite3.connect('./data/1-312-data.db') as sql_conn:
```
This looks right!
